numba/misc/utils.py

```diff
@@ -0,0 +1,65 @@
+from numba.core import cgutils, config, types
+from numba.core.withcontexts import bypass_context
+from numba.core.extending import intrinsic, register_jitable
+import sys
+
+
+# printf
+@intrinsic
+def _printf(typingctx, format_type, *args):
+    """printf that can be called from Numba jit-decorated functions.
+    """
+    if isinstance(format_type, types.StringLiteral):
+        sig = types.void(format_type, types.BaseTuple.from_types(args))
+
+        def codegen(context, builder, signature, fn_args):
+            mod = builder.module
+            value = fn_args[1].value # StructLiteral
+            for i, arg in enumerate(*args):
+                if isinstance(arg, types.StringLiteral):
+                    lv = arg.literal_value
+                    const_str = context.insert_const_string(mod, lv)
+                    value = builder.insert_value(value, const_str, i)
+
+            # Replace %c formatter by %s, since it doesn't work here
+            ft = format_type.literal_value.replace('%c', '%s')
+            cgutils.printf(builder, ft, value)
+        return sig, codegen
+
+
+@register_jitable
+def printf(format_type, *args):
+    if config.DISABLE_JIT:
+        if '%p' in format_type:
```

Comment:
Could this exception be avoided by (untested code follows):
```
new_args = []
def process_format(m):
    orig = m.string[m.start():m.end()]
    if orig == '%p':
        new_args.append(id(args[len(new_args)]))
        return '%#x'
    else:
        new_args.append(args[len(new_args)])
        return orig
format_type = re.sub(r'%[diuxXfeEgGopcs]', process_format, format_type)
args = tuple(new_args)
```
?

Reply:
Unfortunately no. Numba cannot compile this function.

Reply:
The suggestion should live under the `if config.DISABLE_JIT:..` block. Is this still a problem?

Reply:
From what I understand, a `register_jitable` function needs to be compilable in nopython mode. Thus, code inside `if config.DISABLE_JIT` needs to be compilable, even though it will not be executed in nopython mode.